### ISSUE TITLE
gridplot with sizing_mode="stretch_both"/"scale_both" is totally broken since 0.12.11

### ISSUE THREAD

**Issue by cloud-rocket** [bokeh/bokeh#7454]
Charts and toolbar are messed up on the screen. 
```
gridplot(self.panelsGrid,
                        toolbar_location = "right",
                        sizing_mode="stretch_both",
                        toolbar_options=dict(logo=None)
                        )
```

* Tested on Bokeh 0.12.11 and 0.12.13
* Running in `bokeh serve` mode
* Server OS: Debian 9
* Tornado: 4.5.3
* Browser: Chrome, IE 11

I think this MR is breaking it - https://github.com/bokeh/bokeh/pull/7008

Screenshot before browser resizing:
![image](https://user-images.githubusercontent.com/2440892/35332185-c01824ec-0112-11e8-8d61-1f37277616bb.png)

Screenshot after browser resizing
![image](https://user-images.githubusercontent.com/2440892/35332210-d7c55a38-0112-11e8-80d8-2cf5db72a390.png)

**Comment by cloud-rocket** [comment]
Guys, this is a serious bug!
`layout` is working perfectly well, only the gridplot is broken.
I took `examples/howto/layout/dashboard.py` and replaced `layout` with `gridplot` and it looks like this:

![image](https://user-images.githubusercontent.com/2440892/35394067-656e12bc-01ef-11e8-8070-f41d1ed48e20.png)

**Comment by cloud-rocket** [comment]
Any response whatsoever? I provided the way to reproduce it.

Tnx

**Comment by bryevdv** [comment]
@cloud-rocket there is currently *far* more work, than there are people to do it. That mean many things, even important things, will often take a long time. I can't speculate when this will be looked at, the only way to accelerate things is for more people to get involved in contributing.

**Comment by lmmentel** [comment]
I'm getting similar broken behavior when using `column` with Chrome and bokeh 0.12.13

**Comment by cloud-rocket** [comment]
gridplot raises the following errors:
![image](https://user-images.githubusercontent.com/2440892/38316778-2beaf02a-3834-11e8-9c57-26c2606aa51f.png)

Seems like the problem is only when adding external toolbar

**Comment by cloud-rocket** [comment]
@bryevdv,

Looks like CSS `flex` mode you added in MR [7008](https://github.com/bokeh/bokeh/pull/7008/files#diff-4e802fe0944cb0aa4b36e3ad30323c78L3), was not properly added to all elements - when not updated element is added to dashboard, it breaks everything. Probably this is the root problem of multiple recent bug reports with usage of  `scale_width`/`scale_height` and `scale_both`.

I tried manually creating a columns/rows or mixture of them with simple figures - and everything is working fine until I am adding a toolbar. While I am trying to add ToolbarBox with ProxyToolbar (also introduced in 7008), the same way it is done in `gridplot` - it always breaks the whole dashboard. Even when it is a single column (without additional layouts) with toolbar above several figures.

**Comment by helske** [comment]
I see similar issues with Bokeh 0.12.15 and Firefox:
```
import pandas as pd
from bokeh.io import curdoc
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure
from bokeh.layouts import column
from bokeh.models.widgets import Button
from sklearn.datasets import make_blobs
import random

# example data
n_clusters = random.randrange(2, 5, 1)
data, labels = make_blobs(100, 2, n_clusters, random_state=1)
ids = ['id' + str(s + 1) for s in range(100)]
data = pd.DataFrame(data, index=ids, columns=['Variable 1', 'Variable 2'])

source = ColumnDataSource(data)
button = Button(label="Cluster", button_type="success")

sizing_mode = 'scale_both'

plot = figure(plot_width=200, plot_height=200, sizing_mode=sizing_mode)
plot.circle('Variable 1', 'Variable 2', source=source, alpha=0.5)
panels = column([plot], sizing_mode=sizing_mode)

curdoc().add_root(panels)
```
When running bokeh server I don't see the figure at all, but get an error

> Error: dual optimize failed
> solver.js:490
> t.Bokeh</c</t.prototype._dualOptimize
> solver.js:490
> t.Bokeh</c</t.prototype.suggestValue
> solver.js:203
> t.Bokeh</s</t.prototype.suggest_value
> solver.js:83
> t.Bokeh</T</e.prototype.update_constraints
> plot_canvas.js:638
> t.Bokeh</T</e.prototype.render
> plot_canvas.js:689
> t.Bokeh</T</e.prototype._layout
> plot_canvas.js:652
> t.Bokeh</u</e.prototype._layout
> layout_dom.js:185
> t.Bokeh</u</e.prototype._layout
> layout_dom.js:185
> t.Bokeh</u</e.prototype._do_layout
> layout_dom.js:169
> t.Bokeh</u</e.prototype.resize
> layout_dom.js:140
> t.Bokeh</u</e.prototype.connect_signals/<
> layout_dom.js:214

When removing the button I get very small figure (like 50x100 pixels), and when panning the figure, the size of the figure grows with each action.

**Comment by bryevdv** [comment]
Certainly related to several other layout issues

**Comment by bryevdv** [comment]
@mattpap re: discussions with @pzwang it might be good to consider largely decoupling the grid plot from the constraint system as much as possible. Perhaps just the overall grid size can plug in, and the cell sizes can use either default-equal breaks, or user specified percentages.

**Comment by wernight** [comment]
With the toolbal_position='above', adding CSS rule `.bk-button-bar { z-index: 1 }`, makes it work. I didn't try a lot of combinations but it could be an easy short term fix to get 90% there.

**Comment by bryevdv** [comment]
Well, I just realized the follow on code examples were not actually of gridplots as the original issue sites, but AFAIK there was never an MRE provided to test with for the original reported issue.

**Comment by bryevdv** [comment]
OK so as the next best thing I tested the anscombe example with `stretch_both`:

<img width="1264" alt="screen shot 2019-01-30 at 20 45 12" src="https://user-images.githubusercontent.com/1078448/52031175-14aa1880-24d0-11e9-87b9-e36babc08c0c.png">

Also tested with `scale_width` and with a toolbar present. Definitely seems like it can be closed with #8085
